### ISSUE TITLE
Inconsistent documentation for in-place functions at https://pytorch.org/docs/stable/torch.html

### ISSUE THREAD

**Issue by shubho** [pytorch/pytorch#29028]
## 📚 Documentation

In-place functions are documented inconsistently. For example functions that generate samples from a distribution have a section called "In-place random sampling", however there is no other documentation for other in-place functions.

For example `pow_()` and `log_()` are not documented anywhere. Other in-place functions like `fill_()` are referenced in example code but not mentioned.

It would be good to document all in-place functions that are first-class citizens. Or tag functions in some way that have an in-place variant.

**Comment by colesbury** [comment]
The in-place Tensor methods are documented here (although it requires clicking a few links to get to meaningful documentation):

https://pytorch.org/docs/stable/tensors.html#torch.Tensor.log_
https://pytorch.org/docs/stable/tensors.html#torch.Tensor.pow_

There's also `torch.log_` and `torch.pow_` which are not documented and weren't really intended to be a  public API. I think they accidentally got exposed because that was the default in `native_functions.yaml`, but maybe @gchanan can comment on that.

If we really don't want the in-place functions on the torch package than maybe we should deprecate and subsequently remove them.

**Comment by shubho** [comment]
Yeah the question really what is considered "supported and won't go away" and "use at your own risk" - mainly we are trying to support the same in CrypTen - if the in-place functions are deprecated then we won't spend any time with them in CrypTen - also tagging @soumith

**Comment by vadimkantorov** [comment]
Also, there is a variety of inplace interfaces: torch.Tensor.fun_, torch.Tensor.fun(..., out=...), torch.nn.functional.fun(..., inplace = ...) (especially for functions like sigmoid). Not all functions implement all of these which makes it difficult to support optional inplace mode https://github.com/pytorch/pytorch/issues/26644#issuecomment-534762427

**Comment by gchanan** [comment]
@colesbury's answer is correct (although there isn't a `torch.pow_` that I can find).

We'll try to clean this up.  In the meantime, I wouldn't use any `torch.fn_` functions, those are meant to be on the Tensor only.

If you find something that you want to use and there isn't an obviously supported variant of it, open an issue (and tag me) and we'll figure it out.

**Comment by vadimkantorov** [comment]
@gchanan for instance there is torch.Tensor.masked_fill (without an out argument) and inplace torch.Tensor.masked_fill_ - which prevents the inplace mode that you proposed to @SsnL

**Comment by gchanan** [comment]
@vadimkantorov that's a good point, but I think it should be a separate issue.  In my mind there are two problems:
1) The documentation is inconsistent and we expose a bunch of functions that aren't deprecated nor documented so it's hard to know what to use.
2) The actual functions provided aren't consistent so it's difficult to write generic code.

Do you mind filing an issue for the second one?

**Comment by shubho** [comment]
Thanks for the clarifications - @soumith  also DMed that the ones I am looking for are the ones that @colesbury pointed out in https://pytorch.org/docs/stable/tensors.html#torch.Tensor.log - but good that it also uncovered some docs to clean up :)
